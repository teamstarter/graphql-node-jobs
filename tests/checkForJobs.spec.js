const request = require('supertest')
const addMinutes = require('date-fns/addMinutes')

const {
  migrateDatabase,
  seedDatabase,
  getNewServer,
  models,
  closeEverything,
  deleteTables,
  resetDatabase,
} = require('./test-database.js')
const {
  checkForJobs,
  listJobs,
  getNewClient,
  createJob,
  CancelRequestedError,
} = require('./../lib/index')

// This is the maximum amount of time the band of test can run before timing-out
jest.setTimeout(600000)

let server = null
const client = getNewClient(
  `http://localhost:${process.env.PORT || 8080}/graphql`
)

const wait = (time) => new Promise((resolve) => setTimeout(resolve, time))

const acquireJob = (variables) => ({
  query: `mutation($typeList: [String!]!, $workerId: String) {
    acquireJob(
      typeList: $typeList
      workerId: $workerId
    ) {
      id
      name
      status
      output
    }
  }`,
  variables,
  operationName: null,
})

const jobCreate = (variables) => ({
  query: `mutation($job: jobInput!) {
    jobCreate(
      job: $job
    ) {
      id
      name
      status
      output
      jobUniqueId
    }
  }`,
  variables,
  operationName: null,
})

const jobUpdate = (variables) => ({
  query: `mutation($job: jobInput!) {
    jobUpdate(
      job: $job
    ) {
      id
      name
      status
      isUpdateAlreadyCalledWhileCancelRequested
    }
  }`,
  variables,
  operationName: null,
})

const jobList = (variables) => ({
  query: `query($where: SequelizeJSON!) {
    job(
      where: $where
    ) {
      id
      name
      status
      isHighFrequency
    }
  }`,
  variables,
  operationName: null,
})

const customAcquire = (variables) => ({
  query: `mutation($typeList: [String!]!) {
    customAcquire(
      typeList: $typeList
    ) {
      id
    }
  }`,
  variables,
  operationName: null,
})

const recoverJob = (variables) => ({
  query: `mutation($id: Int!) {
    recover(id: $id) {
      id
      status
    }
  }`,
  variables,
  operationName: null,
})

const toggleHoldJobType = (variables) => ({
  query: `mutation($type: String!) {
    toggleHoldJobType(type: $type) {
      id
      type

    }
  }`,
  variables,
  operationName: null,
})

const retryJob = (variables) => ({
  query: `mutation retryJob($id: Int!){
    retryJob(id: $id) {
      id
      name
      type
      status
    }
  }`,
  variables,
  operationName: null,
})

/**
 * Starting the tests
 */
describe('Test the checkForJob function', () => {
  beforeAll(async () => {
    await migrateDatabase()
    await seedDatabase()
    server = await getNewServer()
  })

  beforeEach(async () => {
    await resetDatabase()
  })

  afterEach(async () => {
    await deleteTables()
  })

  afterAll(async (done) => {
    await closeEverything(server, models, done)
  })

  it('checkForJobs allows to simply acquire and update jobs.', async () => {
    const job = await checkForJobs({
      typeList: ['a'],
      client,
      processingFunction: (job) => {
        return { total: 125 }
      },
      looping: false,
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()
  })

  it('checkForJobs allows to simply acquire multiple types of jobs at the same time.', async () => {
    const job = await checkForJobs({
      typeList: ['a', 'b'],
      client,
      processingFunction: (job) => {
        return { total: 125 }
      },
      looping: false,
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()

    const job2 = await checkForJobs({
      typeList: ['a', 'b'],
      client,
      processingFunction: (job) => {
        return { total: 125 }
      },
      looping: false,
    })
    expect(job2).not.toBeUndefined()
    expect(job2).not.toBe(null)
    expect(job2).toMatchSnapshot()
  })

  it('checkForJobs allows to asynchronous processing functions.', async () => {
    const job = await checkForJobs({
      typeList: ['a', 'b'],
      client,
      processingFunction: async (job) => {
        const result = await new Promise((resolve, reject) =>
          setTimeout(() => {
            resolve('plop')
          }, 100)
        )
        return { total: result }
      },
      looping: false,
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()

    const jobEntity = await models.job.findOne({ where: { id: job.id } })
    expect(jobEntity.startedAt).not.toBe(null)
    expect(jobEntity.endedAt).not.toBe(null)
  })

  it('checkForJobs declare jobs as failed when an error is raised.', async () => {
    const job = await checkForJobs({
      typeList: ['a'],
      client,
      processingFunction: async () => {
        const a = {}
        a.awd()
        return a
      },
      looping: false,
    })
    const error = job.output.error.split('\n')[0]
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(error).toMatchSnapshot()

    const jobEntity = await models.job.findOne({ where: { id: 1 } })
    expect(jobEntity.startedAt).not.toBe(null)
    expect(jobEntity.endedAt).not.toBe(null)
    expect(jobEntity.status).toBe('failed')
    expect(jobEntity.output.error.search('at processingFunction')).not.toBe(-1)
  })

  it('checkForJobs processing function can return nothing if needed.', async () => {
    const job = await checkForJobs({
      typeList: ['a'],
      client,
      processingFunction: async () => {},
      looping: false,
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()

    const jobEntity = await models.job.findOne({ where: { id: 1 } })
    expect(jobEntity.startedAt).not.toBe(null)
    expect(jobEntity.endedAt).not.toBe(null)
    expect(jobEntity.status).toBe('successful')
    expect(jobEntity.output).toMatchSnapshot()
  })

  it('The processingFunction expose .', async () => {
    const job = await checkForJobs({
      typeList: ['a'],
      client,
      processingFunction: async (job, { updateProcessingInfo }) => {
        await updateProcessingInfo({ percent: 10 })
      },
      looping: false,
    })
    expect(job).not.toBeUndefined()
    expect(job).not.toBe(null)
    expect(job).toMatchSnapshot()

    const jobEntity = await models.job.findOne({ where: { id: 1 } })
    expect(jobEntity.startedAt).not.toBe(null)
    expect(jobEntity.endedAt).not.toBe(null)
    expect(jobEntity.status).toBe('successful')
    expect(jobEntity.output).toMatchSnapshot()
    expect(jobEntity.processingInfo).toMatchSnapshot()
  })
})
