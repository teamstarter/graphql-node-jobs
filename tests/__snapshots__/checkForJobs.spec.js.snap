// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the checkForJob function The processingFunction expose . 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {},
  "processingInfo": Object {
    "percent": 10,
  },
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the checkForJob function The processingFunction expose . 2`] = `Object {}`;

exports[`Test the checkForJob function The processingFunction expose . 3`] = `
Object {
  "percent": 10,
}
`;

exports[`Test the checkForJob function checkForJobs allows to asynchronous processing functions. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {
    "total": "plop",
  },
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the checkForJob function checkForJobs allows to simply acquire and update jobs. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {
    "total": 125,
  },
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the checkForJob function checkForJobs allows to simply acquire multiple types of jobs at the same time. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {
    "total": 125,
  },
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the checkForJob function checkForJobs allows to simply acquire multiple types of jobs at the same time. 2`] = `
Object {
  "__typename": "job",
  "id": 2,
  "input": Object {},
  "name": "Bla b",
  "output": Object {
    "total": 125,
  },
  "processingInfo": null,
  "status": "successful",
  "type": "b",
}
`;

exports[`Test the checkForJob function checkForJobs declare jobs as failed when an error is raised. 1`] = `"[TypeError: a.awd is not a function] Stack: TypeError: a.awd is not a function"`;

exports[`Test the checkForJob function checkForJobs processing function can return nothing if needed. 1`] = `
Object {
  "__typename": "job",
  "id": 1,
  "input": Object {},
  "name": "Bla a",
  "output": Object {},
  "processingInfo": null,
  "status": "successful",
  "type": "a",
}
`;

exports[`Test the checkForJob function checkForJobs processing function can return nothing if needed. 2`] = `Object {}`;
